### DISCUSSION TITLE
[Feature Request] Fallback URI to OpenURI

### DISCUSSION THREAD

**Discussion by ramcq** [flatpak/xdg-desktop-portal#1250] [Ideas]
A Java application which is hoping to support Flatpak has this code: https://github.com/fedi-to/fedicraft/blob/default/src/main/java/net/fedi_to/fc/os/NonWinMac.java which launches xdg-open if there is a scheme handler available, and launches a fallback URL (helping the user to install whatever is needed, etc) if not.

My first suggestion was to simply remove the check and call xdg-open, expecting an error to be returned in the case there is no handler. However, unlike the unsandboxed xdg-open (which returns 4 in this case), the sandboxed xdg-open wrapper apparently succeeds and returns 0, because the portal (at least here on GNOME) presents a dialog to the user to select or install an app.

Is there some way for the application to distinguish these two cases and guide the user accordingly? Or could we add a fallback URI property to the Open URI method, analagous to https://learn.microsoft.com/en-us/uwp/api/windows.system.launcheroptions.fallbackuri?view=winrt-22621#windows-system-launcheroptions-fallbackuri ?

**Comment by Mikenux** [comment]
Is it about installing specific things? Specific extensions or specific flatpak apps?

For a fallback URI there is this PR: https://github.com/flatpak/xdg-desktop-portal/pull/1203. It concerns https://github.com/flatpak/xdg-desktop-portal/issues/1202. If this applies to mime types, then with this the app can probably instead ask to open a web link if the portal responds there is no scheme?
